lib/Infer/ExhaustiveSynthesis.cpp

```diff
@@ -0,0 +1,567 @@
+//   which logical operations are equivalent at 1 bit? (include icmps in this)
+// add constraints guarding against synthesizing dumb constants
+//   add x, 0; sub x, 0; shift x, 0; mul x, 0; mul x, 1, mul x, 2
+//   shift left by 1
+// aggressively prune dumb instruction sequences
+//   trunc of trunc, sext of sext, zext of zext
+//   trunc to i1 vs. %w = and %v, 1; %x = icmp ne %w, 0
+//   a bloom filter or something
+//   use the solver to recognize non-minimal instructions / instruction sequences?
+// we want to avoid guessing code that's already there on the LHS
+//   hashing?
+// test against CEGIS with LHS components
+// test the width matching stuff
+// take outside uses into account -- only in the cost model?
+// experiment with synthesizing at reduced bitwidth, then expanding the result
+// aggressively avoid calling into the solver
+
+// TODO move this to Inst
+const bool isCmp(Inst::Kind K) {
+  return K == Inst::Eq || K == Inst::Ne || K == Inst::Ult ||
+    K == Inst::Slt || K == Inst::Ule || K == Inst::Sle;
+}
+
+// TODO move this to Inst
+const bool isShift(Inst::Kind K) {
+  return K == Inst::Shl || K == Inst::AShr || K == Inst::LShr;
+}
+
+void hasConstantHelper(Inst *I, std::set<Inst *> &Visited,
+                       std::vector<Inst *> &ConstList) {
+  // FIXME this only works for one constant and keying by name is bad
+  if (I->K == Inst::Var && (I->Name.find("reserved_") != std::string::npos)) {
+    // FIXME use a less stupid sentinel
+    ConstList.push_back(I);
+  } else {
+    if (Visited.insert(I).second)
+      for (auto Op : I->Ops)
+        hasConstantHelper(Op, Visited, ConstList);
+  }
+}
+
+// TODO do this a more efficient way
+// TODO do this a better way, checking for constants by name is dumb
+void hasConstant(Inst *I, std::vector<Inst *> &ConstList) {
+  std::set<Inst *> Visited;
+  hasConstantHelper(I, Visited, ConstList);
+}
+
+// TODO small vector by reference?
+std::vector<Inst *> matchWidth(Inst *I, unsigned NewW, InstContext &IC) {
+  int OldW = isCmp(I->K) ? 1 : I->Width;
+  if (OldW > NewW)
+    return { IC.getInst(Inst::Trunc, NewW, { I }) };
+  if (OldW < NewW)
+    return {
+            IC.getInst(Inst::SExt, NewW, { I }),
+            IC.getInst(Inst::ZExt, NewW, { I }),
+    };
+  return { I };
+}
+
+void addGuess(Inst *RHS, int MaxCost, std::vector<Inst *> &Guesses,
+              int &TooExpensive) {
+  if (souper::cost(RHS) < MaxCost)
+    Guesses.push_back(RHS);
+  else
+    TooExpensive++;
+}
+
+void findVars(Inst *Root, std::vector<Inst *> &Vars) {
+  // breadth-first search
+  std::set<Inst *> Visited;
+  std::queue<Inst *> Q;
+  Q.push(Root);
+  while (!Q.empty()) {
+    Inst *I = Q.front();
+    Q.pop();
+    if (!Visited.insert(I).second)
+      continue;
+    if (I->K == Inst::Var)
+      Vars.push_back(I);
+    for (auto Op : I->Ops)
+      Q.push(Op);
+  }
+}
+
+void getGuesses(std::vector<Inst *>& Guesses,
+                std::vector<Inst* >& Vars,
```

Comment:
sorry, please put the star in the right place

Reply:
also please put the ampersands one position to the right as in  "CandidateMap &CandMap"

Reply:
I'm sorry, I should've been more careful.